### PR TITLE
create alloc_list for accelerated cold-migration

### PR DESCRIPTION
**Pull request by spickartz**
This patch generates a list of allocated memory areas upon request by the hypervisor.

### CODE REVIEW

File: arch/aarch64/mm/memory.c [hunk 1]
```diff
@@ -297,3 +297,9 @@ int memory_init(void)
 oom:
 	return ret;
 }
+
+void *generate_alloc_list(void)
+{
+	//TODO: return mapped memory regions
+	return NULL;
```
Comment (jbreitbart) [inline arch/aarch64/mm/memory.c]: Don't do silent error :wink:

File: arch/x86_64/mm/memory.c [hunk 2]
```diff
@@ -471,3 +472,105 @@ int memory_init(void)
...
+	alloc_list->next = alloc_list->prev = NULL;
+	alloc_list_length++;
+
+	while (next) {
+		/* create new alloc element */
+		free_list_t *new_alloc_elem = (free_list_t *)kmalloc(sizeof(free_list_t));
+		new_alloc_elem->start = cur->end;
+		new_alloc_elem->end = next->start;
+
+		/* append to the list (front) */
```
Comment (jbreitbart) [inline arch/x86_64/mm/memory.c]: I guess you want a function for that?

File: arch/x86_64/mm/memory.c [hunk 3]
```diff
@@ -471,3 +472,105 @@ int memory_init(void)
...
+		new_alloc_elem->next = alloc_list;
+		new_alloc_elem->prev = NULL;
+		alloc_list->prev = new_alloc_elem;
+		alloc_list = new_alloc_elem;
+
+		cur = cur->next;
+		next = cur->next;
+		alloc_list_length++;
```
Comment (jbreitbart) [inline arch/x86_64/mm/memory.c]: Is there any reason why the list itself does not keep track of its size?

File: arch/x86_64/mm/memory.c [hunk 4]
```diff
@@ -471,3 +472,105 @@ int memory_init(void)
...
+		/* append to the list (front) */
+		new_alloc_elem->next = alloc_list;
+		new_alloc_elem->prev = NULL;
+		alloc_list->prev = new_alloc_elem;
+		alloc_list = new_alloc_elem;
+		alloc_list_length++;
+	}
+
+	/* check if we included the IO_GAP */
+	for (cur=alloc_list_last; cur != NULL; cur=cur->prev) {
```
Comment (jbreitbart) [inline arch/x86_64/mm/memory.c]: Why are you not doing this while creating the list in the first place?

File: arch/x86_64/mm/memory.c [hunk 5]
```diff
@@ -471,3 +472,105 @@ int memory_init(void)
...
+			alloc_list_length++;
+			break;
+		}
+	}
+
+	LOG_INFO("generate_alloc_list(): SUCCESS!\n");
+	return (void*)alloc_list_last;
+}
+
+void free_list_rev(void *list)
```
Comment (jbreitbart) [inline arch/x86_64/mm/memory.c]: Is there a free_list_forward? Does it make any sense? Is it important that the freeing is done in reverse order? I would just call the `function free_list`? or `free_free_list`?

Comment (jbreitbart) [reply arch/x86_64/mm/memory.c]: Or Maybe `free_alloc_list`?

File: arch/x86_64/mm/memory.c [hunk 6]
```diff
@@ -471,3 +472,105 @@ int memory_init(void)
...
+			new_elem->prev = cur->prev;
+			new_elem->next = cur;
+			cur->prev = new_elem;
+			alloc_list_length++;
+			break;
+		}
+	}
+
+	LOG_INFO("generate_alloc_list(): SUCCESS!\n");
+	return (void*)alloc_list_last;
```
Comment (jbreitbart) [inline arch/x86_64/mm/memory.c]: Why are you returning a `void*` and not `free_list_t*`?

File: arch/x86_64/mm/memory.c [hunk 7]
```diff
@@ -471,3 +472,105 @@ int memory_init(void)
...
+	}
+
+	/* allocation behind last free region */
+	if (limit != cur->end) {
+		/* create new alloc element */
+		free_list_t *new_alloc_elem = (free_list_t *)kmalloc(sizeof(free_list_t));
+		new_alloc_elem->start = cur->end;
+		new_alloc_elem->end = limit - cur->end;
+
+		/* append to the list (front) */
```
Comment (jbreitbart) [inline arch/x86_64/mm/memory.c]: So, how about a function? :stuck_out_tongue:

**Comment by spickartz** [comment]
Closed. Using another approach now (see #101).

**Pull request closed without merging**